File: backend_samples/library/javascript/AzureAIVisionFaceDeviceAttestation/src/ios/cbor.ts
import { Decoder } from 'cbor';
import type { CborValue } from './types';

/** Decode one App Attest CBOR item and return its ending offset. */
export function decodeCbor(buf: Buffer, pos = 0, depth = 0): [CborValue, number] {
  if (!Number.isInteger(pos) || pos < 0 || pos >= buf.length) throw new Error('CBOR: invalid offset');
  if (depth > 16) throw new Error('CBOR: max nesting depth exceeded');
  const options = { preferMap: true, max_depth: 16 - depth, extendedResults: true };
  const decoder = new Decoder(options);
  let result: { value: unknown; length: number } | undefined;
  decoder.on('error', () => {});
  decoder.on('start', (major: number, length: unknown) => {
    if (major === 6 || typeof length === 'symbol') throw new Error('CBOR: tags and indefinite encodings are unsupported');
  });
  decoder.on('more-bytes', (major: number) => {
    if (major === 7) throw new Error('CBOR: simple values and floats are unsupported');
  });
  decoder.on('value', validateValue);
  decoder.on('data', value => {
    result = value;
    decoder.close();
  });
  try {
    decoder.write(buf.subarray(pos));
    if (!result) throw new Error('CBOR: malformed or unsupported input');
    return [validateValue(result.value), pos + result.length];
  } finally {
    decoder.destroy();
  }
}

function validateValue(value: unknown): CborValue {
  if (typeof value === 'string' || Buffer.isBuffer(value)) return value;
  if (typeof value === 'number' && Number.isSafeInteger(value)) return value;
  if (Array.isArray(value)) return value.map(validateValue);
  if (value instanceof Map) {
    for (const [key, item] of value) {
      if (typeof key !== 'string' && (typeof key !== 'number' || !Number.isSafeInteger(key))) {
        throw new Error('CBOR: unsupported map key type');
      }
      validateValue(item);
    }
    return value;
  }
  throw new Error('CBOR: unsupported value');
}

File: backend_samples/library/dotnet/AzureAIVisionFaceDeviceAttestation/src/Ios/Cbor.cs
using System.Formats.Cbor;

namespace Azure.AI.Vision.Face.DeviceAttestation.Ios;

/// <summary>
/// CBOR adapter for Apple App Attest payloads. Supports unsigned
/// int, negative int, byte string, text string, array, and map. Indefinite
/// lengths, tags, and floats are rejected so malformed input fails loudly.
/// Values are boxed as: long, string, byte[], List&lt;object?&gt;, or
/// Dictionary&lt;object, object?&gt; (string/long keys).
/// </summary>
internal static class Cbor
{
    public static (object? Value, int NextPos) Decode(byte[] buf, int pos = 0)
    {
        var reader = new CborReader(buf.AsMemory(pos), CborConformanceMode.Lax, allowMultipleRootLevelValues: true);
        var value = ReadValue(reader, 0);
        return (value, buf.Length - reader.BytesRemaining);
    }

    private static object ReadValue(CborReader reader, int depth)
    {
        if (depth > 16) throw new FormatException("CBOR: max nesting depth exceeded");
        switch (reader.PeekState())
        {
            case CborReaderState.UnsignedInteger:
            case CborReaderState.NegativeInteger:
                return reader.ReadInt64();
            case CborReaderState.ByteString:
                return reader.ReadByteString();
            case CborReaderState.TextString:
                return reader.ReadTextString();
            case CborReaderState.StartArray:
            {
                int length = reader.ReadStartArray() ?? throw new FormatException("CBOR: indefinite array");
                var arr = new List<object?>();
                for (int index = 0; index < length; index++)
                    arr.Add(ReadValue(reader, depth + 1));
                reader.ReadEndArray();
                return arr;
            }
            case CborReaderState.StartMap:
            {
                int length = reader.ReadStartMap() ?? throw new FormatException("CBOR: indefinite map");
                var map = new Dictionary<object, object?>();
                for (int index = 0; index < length; index++)
                {
                    var key = ReadValue(reader, depth + 1);
                    if (key is not string && key is not long)
                        throw new FormatException("CBOR: unsupported map key type");
                    map[key] = ReadValue(reader, depth + 1);
                }
                reader.ReadEndMap();
                return map;
            }
            default:
                throw new FormatException($"CBOR: unsupported state {reader.PeekState()}");
        }
    }
}
